nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/processors/script/ExecuteScript.java

```diff
@@ -0,0 +1,585 @@
+                        modulePath = oldValue;
+                        logger.error(t.getLocalizedMessage(), t);
+                    }
+
+                } else if (SCRIPT_ENGINE.equals(descriptor)) {
+                    // The script engine has changed, so we need to set up a new instance for the selected
+                    // engine name
+                    scriptEngineName = newValue;
+                    setupEngine();
+                }
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Creates a classloader to be used by the selected script engine and the provided script file. This
+     * classloader has the ExecuteScript's classloader as a parent (versus the current thread's context
+     * classloader) and also adds the specified module directory to the classpath. This enables scripts
+     * to use other scripts, modules, etc. without having to build them into the ExecuteScript NAR.
+     * If the parameter is null or empty, ExecuteScript's classloader is returned
+     *
+     * @param modulePath The path to a directory containing modules to be used by the script(s)
+     */
+    private ClassLoader createScriptEngineModuleClassLoader(String modulePath) {
+        URLClassLoader newModuleClassLoader = null;
+        if (StringUtils.isEmpty(modulePath)) {
+            return ExecuteScript.class.getClassLoader();
+        }
+        try {
+            newModuleClassLoader =
+                    new URLClassLoader(
+                            new URL[]{new File(modulePath).toURI().toURL()}, ExecuteScript.class.getClassLoader());
+        } catch (MalformedURLException mue) {
+            getLogger().error("Couldn't find modules directory at " + modulePath, mue);
+        }
+        return newModuleClassLoader;
+    }
+
+    /**
+     * Evaluates the given script body (or file) using the current session, context, and flowfile. The script
+     * evaluation expects a FlowFile to be returned, in which case it will route the FlowFile to success. If a script
+     * error occurs, the original FlowFile will be routed to failure. If the script succeeds but does not return a
+     * FlowFile, the original FlowFile will be routed to no-flowfile
+     *
+     * @param context the current process context
+     * @param session the current process session
+     * @throws ProcessException if any error occurs during script evaluation
+     */
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        synchronized (isInitialized) {
+            if (!isInitialized.get()) {
+                createResources();
+            }
+        }
+        ProcessorLog log = getLogger();
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String[] args = context.getProperty(SCRIPT_ARGS).evaluateAttributeExpressions().getValue().split(";");
```

Comment:
should probably be calling evaluateAttributeExpressions(flowFile) here instead of just evaluateAttributeExpressions() - as-is, it will not evaluate FlowFile Attributes.

Also, rather than splitting by semi-colon, consider splitting by spaces and allow double-quotes to be used to encapsulate spaces. This is a much more natural way of specifying arguments, as this is how it works on command-line. Code exists already in the standard-processors module for this - org.apache.nifi.processors.standard.util.ArgumentUtils - perhaps that could be pulled out into the processor-utils module?